### PR TITLE
Support mode 'nearest-exact' for torch.nn.functional.interpolate

### PR DESCRIPTION
**Pull request by lianakoleva**
First step towards addressing #1941, referencing [ATen source](https://github.com/pytorch/pytorch/commit/6adbe044e39c8e8db158d91e151aa6dead6e9aa4).

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -5098,7 +5098,7 @@ def _interpolate_scale_factor_helper(
     scale_factor: Sequence[float] | float,
     mode: str = "nearest",
 ) -> TensorLike:
-    assert mode == "nearest"
+    assert mode == "nearest" or mode == "nearest-exact"
```
Comment (mruberry) [inline thunder/torch/__init__.py:5101]: While you're here, would you please update this to throw a `ValueError`:

```
if mode not in ('nearest', 'nearest-exact'):
  raise ValueError(f"_interpolate_scale_factor_helper expected mode to be 'nearest' or 'nearest-exact', but got {mode}")
```

File: thunder/torch/__init__.py [hunk 2]
```diff
@@ -5119,13 +5119,15 @@ def _interpolate_scale_factor_helper(
         )
 
     # perform nearest up/down-sampling
-    def nearest_sampler(t, input_dim, output_dim, *, scale, dim):
+    def nearest_sampler(t, input_dim, output_dim, *, scale, dim, exact):
```
Comment (mruberry) [inline thunder/torch/__init__.py:5122]: Would you add type annotations to this function? I think they would greatly improve its readability

File: thunder/torch/__init__.py [hunk 3]
```diff
@@ -5147,6 +5149,7 @@ def dim_expander(t, dim, n_repeats):
         # dimenions corresponding to batches and channels.
         curr_dim = 2 + (len(spatial_dims) - k - 1)
 
+        exact = mode == "nearest-exact"
```
Comment (mruberry) [inline thunder/torch/__init__.py:5152]: nit: 

```
exact: bool = (mode == 'nearest-exact')
```

File: thunder/torch/__init__.py [hunk 4]
```diff
@@ -5210,8 +5213,8 @@ def interpolate(
     antialias: bool = False,
 ) -> TensorLike:
     utils.check(
-        mode == "nearest",
-        lambda: f"only mode='nearest' is supported at the moment, but got {mode=}",
+        (mode == "nearest" or mode == "nearest-exact"),
```
Comment (mruberry) [inline thunder/torch/__init__.py:5216]: Nice update

**Comment by mruberry** [comment]
Hey @lianakoleva! Cool stuff. I made a few small comments. @beverlylytle, would you also like to take a look?

To test the implementation, we'll need to extend the "sample inputs" of `interpolate`, which are defined here:

https://github.com/Lightning-AI/lightning-thunder/blob/4acf8198cfed1761528f9d10d2bc1a5356cf9ce9/thunder/tests/opinfos.py#L9320

In particular, I think this part:

https://github.com/Lightning-AI/lightning-thunder/blob/4acf8198cfed1761528f9d10d2bc1a5356cf9ce9/thunder/tests/opinfos.py#L9363

could be edited to look like:

```
yield SampleInput(make(a_shape), size=size)
yield SampleInput(make(a_shape), size=size, mode='nearest-exact')
```

So that both modes are tested. Please let us know if you have any questions about this, as the testing can be confusing at first!

File: thunder/torch/__init__.py [hunk 5]
```diff
@@ -5119,13 +5119,15 @@ def _interpolate_scale_factor_helper(
...
         # It is expected that output_dim = int(input_dim * scale).
         # Indices [0, ..., output_dim - 1] are mapped to [0, ..., input_dim - 1]
-        # with the rule i -> int(i * scale)
+        # with the rule i -> int(i * scale) or i -> round((i + 0.5) * scale - 0.5),
+        # corresponding to modes 'nearest' or 'nearest-exact', respectively.
         # Values at these indices is the result.
+        # References https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/UpSample.h
         selected_idx = arange(0, output_dim, device=a.device)
-        selected_idx = to(selected_idx * scale, selected_idx.dtype)
+        selected_idx = to((selected_idx + exact * 0.5) * scale, selected_idx.dtype)
```
[the remark this answers is not in the document]
Comment (lianakoleva) [reply thunder/torch/__init__.py:5130]: In the case provided here, `to(a, dtype)` function calls `clang.maybe_convert_to_dtype(a, dtype)` ([link](https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/torch/__init__.py#L491)), which calls `prims.convert_element_type(a, dtype)` ([link](https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/clang/__init__.py#L174)), which calls `_convert_element_type_meta` ([link](https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/core/prims.py#L1907)), which calls `dtype(a)` ([link](https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/core/prims.py#L1898)). I assumed `dtype(a)` is a Python casting. Python `int` casting [truncates floats towards 0](https://docs.python.org/3/library/functions.html#int); for natural numbers (range of inputs for `nearest_sampler`), this is a floor operation. For a floor operation, I think we are good to drop the `-0.5`. If we need to look at any other dtype combinations or anything looks wrong here, let me know.

[the remark this answers is not in the document]
Comment (beverlylytle) [reply thunder/torch/__init__.py:5141]: Good, but I'm still a little confused.  Why is this the right place to add the 0.5 factor and not where `output_dim` is defined in line 5139/5152?

In `_interpolate_size_helper`, it looks like the mode is being ignored.  The tests are passing, but it looks like test coverage for `interpolate` isn't particularly refined and the subtle difference between 'nearest' and 'nearest-exact' is falling through the cracks.  Considering the impact to CI test duration, do you think it's worthwhile to add a test case where the result of `interpolate` will differ based on the mode?

[the remark this answers is not in the document]
Comment (lantiga) [reply thunder/torch/__init__.py:5141]: @lianakoleva I agree with @beverlylytle, it'd be great if you could add a test case where the difference between the two methods is exercised

[the remark this answers is not in the document]
Comment (lianakoleva) [reply thunder/torch/__init__.py:5321]: done

**Review by lantiga — APPROVED** [review]

File: thunder/torch/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (beverlylytle) [reply thunder/torch/__init__.py:5321]: I'm still confused.  This line is out of field for commenting, but in:
```
def _interpolate_size_helper(
    a: TensorLike,
    size: Sequence[int] | int,
    mode: str = "nearest",
) -> TensorLike:
    batch, channels, *spatial_dims = a.shape
    dim = len(spatial_dims)
    if isinstance(size, (int, IntegerProxy)):
        utils.check(size > 0, lambda: f"{size=} is expected to be greater than zero")
        size = (size,) * dim
    else:
        utils.check(
            (
                isinstance(size, Sequence)
                and len(size) == dim
                and all(isinstance(s, (int, IntegerProxy)) and s > 0 for s in size)
            ),
            lambda: f"{size=} is expected to be a greater than zero integer "
            f"or a sequence of strictly positive integers of length {dim}",
        )
    scale_factor = tuple(output_size / input_size for output_size, input_size in zip(size, spatial_dims))
    return _interpolate_scale_factor_helper(a, scale_factor)
``

shouldn't the last line be `return _interpolate_scale_factor_helper(a, scale_factor, mode)`?

[the remark this answers is not in the document]
Comment (beverlylytle) [reply thunder/torch/__init__.py:5321]: I guess it isn't caught by the new test because `scale_factor` is specified and not `size`.
